File: solutions/php/largest-series-product/1/LargestSeriesProduct.php
<?php
declare(strict_types=1);

class Series
{
    private string $input;

    function __construct(string $input)
    {
        $this->input = $input;
    }

    function largestProduct(int $span): int
    {
        if ($span < 0) {
            throw new \InvalidArgumentException('Span must not be negative');
        }
        if ($span > strlen($this->input)) {
            throw new \InvalidArgumentException(
                'Span must be smaller than string length',
            );
        }
        if (preg_match('/[^0-9]/', $this->input)) {
            throw new \InvalidArgumentException(
                'Digits input must only contain digits',
            );
        }

        if ($span === 0) {
            return 1;
        }

        $maxProduct = 0;

        for ($i = 0; $i <= strlen($this->input) - $span; $i++) {
            $series = substr($this->input, $i, $span);
            $product = array_product(str_split($series));
            if ($product > $maxProduct) {
                $maxProduct = $product;
            }
        }

        return $maxProduct;
    }
}

File: solutions/go/largest-series-product/2/largest_series_product.go
package lsproduct

import "errors"

func LargestSeriesProduct(digits string, span int) (int64, error) {
	if span < 0 {
		return 0, errors.New("span must not be negative")
	}
	if span == 0 {
		return 1, nil
	}

	digitsLength := len(digits)
	if span > digitsLength {
		return 0, errors.New("span must be smaller than string length")
	}

	numbers := make([]int64, digitsLength)
	for i, c := range digits {
		if c < '0' || c > '9' {
			return 0, errors.New("digits input must only contain digits")
		}
		numbers[i] = int64(c - '0')
	}

	maxProduct := int64(0)

	for i := 0; i <= digitsLength-span; i++ {
		product := int64(1)
		for j := i; j < i+span; j++ {
			product *= numbers[j]
		}
		if product > maxProduct {
			maxProduct = product
		}
	}
	return maxProduct, nil
}
